File: zZ/java/merge-k-sorted-lists.java
/**
 * Merge k Sorted Lists
 *  
 * Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.
 */

public class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
    }
}

public class Solution {
    private static class ListNodeComparator implements Comparator<ListNode> {
        @Override
        public int compare(ListNode n1, ListNode n2) {
            return n1.val - n2.val;
        }
    }

    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null) {
            throw new NullPointerException();
        }

        Queue<ListNode> priorityQueue = new PriorityQueue<>(new ListNodeComparator());
        for (ListNode list : lists) {
            if (list != null) {
                priorityQueue.add(list);
            }
        }

        ListNode precedingHeadPointer = new ListNode(0);
        ListNode curr = precedingHeadPointer;
        while (!priorityQueue.isEmpty()) {
            ListNode next = priorityQueue.poll();
            curr.next = next;
            curr = curr.next;
            if (next.next != null) {
                priorityQueue.add(next.next);
            }
        }

        return precedingHeadPointer.next;
    }
}

File: zZ/javascript/leetcode/merge-k-sorted-lists.js
// 23. Merge k Sorted Lists

// Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.

// Example:

// Input:
// [
//   1->4->5,
//   1->3->4,
//   2->6
// ]
// Output: 1->1->2->3->4->4->5->6

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
    if (!isArray(lists)) {
        throw "invalid input types"
    }
    
    return mergeKListsHelper(lists, 0, lists.length - 1)
};

function isArray(item) {
    return Array.isArray(item)
}

function mergeKListsHelper(lists, leftIndex, rightIndex) {
    if (leftIndex > rightIndex) {
        return null
    }
    
    if (leftIndex === rightIndex) {
        return lists[leftIndex]
    }
    
    const middleIndex = leftIndex + Math.trunc((rightIndex - leftIndex) / 2)
    
    return mergeTwoLists(
        mergeKListsHelper(lists, leftIndex, middleIndex),
        mergeKListsHelper(lists, middleIndex + 1, rightIndex)
    )
}

function mergeTwoLists(list1, list2) {
    const dummy = new ListNode(-1)
    let curr = dummy
    
    while (list1 && list2) {
        if (list1.val <= list2.val) {
            curr.next = list1
            list1 = list1.next
        } else {
            curr.next = list2
            list2 = list2.next
        }
        
        curr = curr.next
    }
    
    curr.next = list1 ? list1 : list2
    return dummy.next
}